src/gpl/src/mbff.h

```diff
@@ -0,0 +1,119 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+///////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include <vector>
+#include "odb/db.h"
+#include "utl/Logger.h"
+
+namespace utl {
+class Logger;
+}
+
+namespace gpl {
+
+struct Tray {
+    odb::Point pt;
+    std::vector<odb::Point> slots;
+    std::vector<int> cand = std::vector<int>(1, -1);
+};
+
+struct Flop {
+    odb::Point pt;
+    int idx = 0;
+    float prob = 0;
+
+    bool operator<(Flop &a) { return prob < a.prob; }
+};
+
+struct Path {
+    int a = 0;
+    int b = 0;
+};
+
+class MBFF {
+
+  public:
+    MBFF(int num_flops, int num_paths, const std::vector<float> &x,
+         const std::vector<float> &y,
+         const std::vector<std::pair<int, int> > &paths, int threads,
+         utl::Logger *logger);
+    ~MBFF();
+
+    void Run(int mx_sz, float alpha, float beta);
+
+  private:
+    int GetRows(const int slot_cnt);
+    int GetBitCnt(const int bit_idx);
+    float GetDist(const odb::Point &a, const odb::Point &b);
+
+    std::vector<odb::Point> GetSlots(const odb::Point &tray, const int rows,
+                                     const int cols);
+    Flop GetNewFlop(const std::vector<Flop> &prob_dist, const float tot_dist);
+    std::vector<Tray> GetStartTrays(std::vector<Flop> flops,
+                                    const int num_trays, const float AR);
+    Tray GetOneBit(const odb::Point &pt);
+
+    std::vector<std::pair<int, int> >
+    MinCostFlow(const std::vector<Flop> &flops, std::vector<Tray> &trays,
+                int sz);
+
+    float GetSilh(const std::vector<Flop> &flops,
+                  const std::vector<Tray> &trays,
+                  const std::vector<std::pair<int, int> > &clusters);
+
+    std::vector<std::vector<Flop> > KMeans(const std::vector<Flop> &flops,
+                                           int K);
+    std::vector<std::vector<Flop> > KMeansDecomp(const std::vector<Flop> &flops,
+                                                 int MAX_SZ);
+
+    std::vector<std::pair<int, int> >
+    RunCapacitatedKMeans(const std::vector<Flop> &flops,
+                         std::vector<Tray> &trays, int sz, int iter);
+
+    std::vector<std::vector<Tray> > RunSilh(const std::vector<Flop> &pointset);
```

Comment:
Most, if not all, of these should be returning references to avoid high-overhead copying.

Reply:
would it instead be better to have all of these functions return void, and have all the functions modify a vector passed in by reference?

Reply:
I don't think you need vectors passed in by reference. Returning a reference would be more readable and solve the copy issue.

Reply:
@therealshreyas I'm sorry, I thought you were returning class members in this function, not a vector created inside the function. The approach you suggested is better for this and potentially other cases in your code.